File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        def subsetsum(trgt):
            ln=len(nums)
            dp = [[False]*(trgt+1) for _ in range(ln+1)]
            for l in range(ln+1):
                for t in range(trgt+1):
                    if t==0:
                        dp[l][t]=True
                    elif l==0:
                        dp[l][t]=False
                    elif t>=nums[l-1]:
                        dp[l][t] = dp[l-1][t] or dp[l-1][t-nums[l-1]]
                    else:
                        dp[l][t] = dp[l-1][t]
            return dp[ln][trgt]
        
        sumt=sum(nums)
        return False if sumt%2 else subsetsum(sumt//2)
                  

# class Solution:
#     def canPartition(self, nums: List[int]) -> bool:
#         target = sum(nums)
#         if target%2==1:
#             return False
#         target = target//2
        
#         def helper(ptr, target):
#             if target==0:
#                 return True
#             if ptr>=len(nums) or target<0:
#                 return False
#             if memo[ptr][target]==-1:
#                 memo[ptr][target] = int(helper(ptr+1, target) or helper(ptr+1, target-nums[ptr]))
#             return memo[ptr][target]==1
            
#         memo = [[-1]*(target+1) for _ in range(len(nums))]
#         return helper(0, target)
        

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int target=0;
        for(int n:nums) target += n;
        return target%2==0 ? subsetsum(nums, target/2) : false;
    }
    boolean subsetsum(int[] nums, int trgt){
        int lngt = nums.length;
        boolean dp[][] = new boolean[lngt+1][trgt+1];
        for(int l=0; l<=lngt; ++l){
            for(int t=0; t<=trgt; ++t){
                if(t==0) dp[l][t] = true;
                else if(l!=0)
                    dp[l][t] = t>=nums[l-1] ? dp[l-1][t] || dp[l-1][t-nums[l-1]] : dp[l-1][t];
            }
        }
        return dp[lngt][trgt];
    }
}


// class Solution {
//     public boolean canPartition(int[] nums) {
//         int target=0;
//         for(int n:nums){
//             target += n;
//         }
//         if(target%2==1){
//             return false;
//         }
//         target /= 2;
//         boolean[][] memo = new boolean[target+1][nums.length];
//         for(int t=0; t<=target; ++t){
//             for(int ptr=0; ptr<nums.length; ++ptr){
//                 if(t==0){
//                     memo[t][ptr]=true;
//                 }else if(ptr==0){
//                     memo[t][ptr]=false;
//                 }else{
//                     memo[t][ptr]=memo[t][ptr-1];
//                     if(t>=nums[ptr] && !memo[t][ptr]){
//                         memo[t][ptr] = memo[t-nums[ptr]][ptr-1];
//                     }
//                 }
//             }
//         }
//         return memo[target][nums.length-1];
//     }
// }


// class Solution {
//     int[][] memo;
//     public boolean canPartition(int[] nums) {
//         int target=0;
//         for(int n:nums){
//             target += n;
//         }
//         if(target%2==1){
//             return false;
//         }
//         target /= 2;
//         memo = new int[target+1][nums.length+1];
//         return knap(0, nums, target);
//     }
//     boolean knap(int ptr, int[] nums, int target){
//         if(target==0){
//             return true;
//         }else if(ptr>=nums.length || target<0){
//             return false;
//         }else if(memo[target][ptr] != 0){
//             return memo[target][ptr]==1 ? true : false; 
//         }else{
//             boolean res = knap(ptr+1, nums, target) || knap(ptr+1, nums, target-nums[ptr]);
//             memo[target][ptr] = (res) ? 1 : -1;
//             return res;
//         }
//     }
// }

// class Solution {
//     public boolean canPartition(int[] nums) {
//         int sumt=0;
//         for(int n:nums){
//             sumt+=n;
//         }
//         int target=sumt/2;
//         return knap(0, nums, target);
//     }
//     boolean knap(int ptr, int[] nums, int target){
//         if(target==0){
//             return true;
//         }else if(ptr>=nums.length || target<0){
//             return false;
//         }else{
//             return knap(ptr+1, nums, target) || knap(ptr+1, nums, target-nums[ptr]);
//         }
//     }
// }